Create Usage section in documentation + API (#1577)
dottxt-ai/outlines: docs_new/features/core/output_types.md

---
title: Output Types
---

# Output Types

Outlines provides a simple and intutive way of defining the output format of text generation.

## Overview

The general idea is that you should provide as an output type what you would give as the type hint of the return type of a function.

> Thus, you can use most Python native types as well as the additional types from the `typing` module and some widespread Python packages such as `pydantic`.

Reviewer: Seems we also have this stuff down below in the "structured output" section. We should move that up top to give people a quick overview of available types.

> Thus, you can use most Python native types as well as the additional types from the `typing` module and some widespread Python packages such as `pydantic`.

Author: Unsupported types: co,plex, rqnge, set, frozenset, bytes, bytearray and memoryview

### Basic Types

The most straightfoward form of structed generation is to return an answer that conforms to a given basic type such as an int or a python list. You can use the basic Python types, but also the types contained in the `typing` library and combine them for more complex use cases. Some examples of valid Outlines output types:

```python
from typing import Dict, List, Optional, Tuple, Union

output_type = float
output_type = Dict[int, str]
output_type = Tuple[Union[str, Dict[int, float]], List[bool], Optional[int]]
```

### Multiple Choice

You can use your model to do classification by using the `Literal` or `Enum` output types. For instance:

```python
from typing import Literal
from enum import Enum

class PizzaOrBurger(Enum):
    pizza = "pizza
    burger = "burger

# Equivalent multiple-choice output types
output_type = Literal["pizza", "burger"]
output_type = PizzaOrBurger
```

> 

Author: Yes, I think it makes more sense in the guide section

### Structured Objects

What we call structured object here are roughly all common types that can be turned into a JSON schema. That includes:
- A Pydantic class
- A Dataclass
- A TypedDict
- A GenSon SchemaBuilder
- A Callable (the parameters are turned into the keys and the type hinting is used to define the types of the values)

For instance:

```python
from dataclasses import dataclass

@dataclass
class Character:
    name: str
    age: int

ouput_type = Character

def character(name: str, age: int):
    return None

output_type = character
```

> On top of those five types that can directly be used with a model, there are also two other json schema formats for which we need to use an Outlines-specific class: JSON schema strings and dicts. As those do not have a distinct type, we must wrap them into an `outlines.types.JsonSchema` object. For instance:

Reviewer: ```suggestion
There are two other JSON schema formats that require Outlines-specific classes: JSON schema strings and dictionaries. 

The desired output type of schema strings and dictionaries is ambiguous, so Outlines requires that these be wrapped in a `outlines.types.JsonSchema` object. For instance:
```

```python
from outlines.types import JsonSchema

schema_string = '{"type": "object", "properties": {"answer": {"type": "number"}}}'
output_type = JsonSchema(schema_string)

schema_dict = {
    "type": "object",
    "properties": {
        "answer": {"type": "number"}
    }
}
output_type = JsonSchema(schema_dict)
```

### Regex Patterns

> You can ask your model to generate text that follows a regex pattern. As the latter are expressed as simple raw string literals, we need to use an Outlines-specific class. Thus, the regex string should be be wrapped in an `outlines.types.Regex` object. For instance:

Reviewer: ```suggestion
Outlines provide support for text generation constrained by regular expressions. Since regular expressions are expressed as simple raw string literals, regex strings must wrapped in an `outlines.types.Regex` object to clarify the expected return type.
```

```python
from outlines.types import Regex

regex = r"[0-9]{3}"
output_type = Regex(regex)
```

> The `types` module contains a few common regex patterns stored in variable you can import and directly use as output types. There are for instance the patterns for a sentence, an email address and an isbn reference. For instance:

Reviewer: ```suggestion
The `outlines.types` module contains a few common regex patterns stored in variable you can import and directly use as output types. Common patterns include a sentence, an email address and an [ISBN reference](https://en.wikipedia.org/wiki/ISBN). For instance:
```

```python
from outlines.types import sentence

print(type(sentence)) # outlines.types.dsl.Regex
print(sentence.pattern) # [A-Z].*\s*[.!?]
```

> To help you create complex regex patterns yourself, you can rely on the Outlines [regex DSL]().

Reviewer: ```suggestion
To help you create complex regex patterns yourself, you can use the Outlines [regex DSL](../../utility/regex_dsl).
```

### Context-Free Grammars

> Outlines allows you to generate text that respects the syntax of a context-free grammar. For that, you need to define a Lark grammar. As the latter is expressed as a string, we need to use an Outlines-specific class. Thus, the CFG string should be be wrapped in an `outlines.types.CFG` object. For instance:

Reviewer: ```suggestion
Outlines allows you to generate text that respects the syntax of a context-free grammar. Context-free grammars are defined using [Lark](https://lark-parser.readthedocs.io/en/latest/index.html), a grammar language. Since grammars are expressed as a string, Large CFG strings should be be wrapped in an `outlines.types.CFG` object. For instance:
```

```python
from outlines.types import CFG

grammar_string = """
    start: expr
    expr: "{" expr "}" | "[" expr "]" |
"""
output_type = CFG(grammar_string)
```

> You can find a few Lark grammar examples in the [grammars module]().

Reviewer: I'm not sure what to do about links to things we don't have yet. Should this link to the API reference or to somewhere in the user guide?

Author: To me it makes sense to point to the API reference for the moment and when guides are added we update the link

Reviewer: Works for me

## Output type availability

> The output types presented above are not available for all models as some have only limited support for structured outputs. Please refer to the documentation of the specific model you wish to use to know what output types it supports.

Reviewer: We should probably add a `models/index.md` file in the root of the `models/` dir, and then update the mkdocs tree to point to that index file. See how `index.md` is included directly under the name [here](https://github.com/dottxt-ai/outlines/blob/b6ffccc5581aa920e2af76a8d10ea8345ce86df6/mkdocs.yml#L149-L153)).

That index file should contain hyperlinks to each of the available models, and eventually, a table listing supported features. We can defer the table to a much later date.
